python_ta/transforms/type_inference_visitor.py

```diff
@@ -290,6 +332,35 @@ def visit_ifexp(self, node):
         else:
             node.type_constraints = TypeInfo(Any)
 
+    def visit_comprehension(self, node):
+        arg_type = self.type_constraints.lookup_concrete(node.iter.type_constraints.type)
+        rtype = self._handle_call(node, '__iter__', arg_type).type
+        if isinstance(node.target, astroid.Tuple):
+            for target_node in node.target.elts:
+                target_tvar = node.parent.type_environment.lookup_in_env(target_node.name)
+                self.type_constraints.unify(target_tvar, rtype.__args__[0])
+        else:
+            target_tvar = node.parent.type_environment.lookup_in_env(node.target.name)
+            self.type_constraints.unify(target_tvar, rtype.__args__[0])
+        node.type_constraints = TypeInfo(NoType)
+
+    def visit_listcomp(self, node):
+        # based on the types of the generators.. can set up the type constraints.
+        # the type of the list comprehension should be a list of the elt's types
+        # if the elt is a name node, look up the name in the listcomp node's env and set the type const as that
+        # if it is an operation node,
+        node.type_constraints = TypeInfo(List[node.elt.type_constraints.type])
+
+    def visit_dictcomp(self, node):
+        # TODO: name node stored in .key attribute is visited last, thus we must do a lookup.
```

Comment:
There is no reason to treat the `key` as special. You should do `lookup_concrete` for all of these types.